src/main/scala/tech/sourced/gemini/Gemini.scala

```diff
@@ -81,39 +83,120 @@ class Gemini(session: SparkSession, log: Slf4jLogger, keyspace: String = Gemini.
   def query(inPath: String,
             conn: Session,
             bblfshClient: BblfshClient,
-            feClient: FeatureExtractor): Iterable[RepoFile] = {
+            feClient: FeatureExtractor,
+            docFreqPath: String = "",
+            paramsFilePath: String = "", // TODO remove when we implement hash
+            htNum: Int = 0,
+            bandSize: Int = 0): (Iterable[RepoFile], Iterable[RepoFile]) = {
     val path = new File(inPath)
     if (path.isDirectory) {
-      findDuplicateProjects(path, conn, keyspace)
       //TODO: implement based on Apollo
-      //findSimilarProjects(path)
+      (findDuplicateProjects(path, conn, keyspace), findSimilarProjects(path))
     } else {
-      findSimilarForFile(path, bblfshClient, feClient)
-      findDuplicateItemForFile(path, conn, keyspace)
+      val duplicates = findDuplicateItemForFile(path, conn, keyspace)
+      var similarShas = Array[String]()
+
+      if (docFreqPath != "" && paramsFilePath != "") {
+        similarShas =
+          findSimilarForFile(path, conn, bblfshClient, feClient, docFreqPath, paramsFilePath, htNum, bandSize)
+          .getOrElse(Array[String]())
+      } else {
+        log.warn("Document frequency or parameters for weighted min hash wasn't provided. Skip similarity query")
+      }
+
+      // value for sha1 in hashtables is path@sha1, but we need only sha1
+      similarShas = similarShas.map(_.split("@")(1))
+      // remove duplicates from similar
+      if (similarShas.isEmpty) {
+        log.info("No similar sha1s")
+      } else {
+        val duplicatedShas = duplicates.map(_.sha).toArray
+        similarShas = similarShas.filterNot(sha => duplicatedShas.contains(sha))
+      }
+
+      val similar = similarShas.flatMap(sha1 => findDuplicateItemForBlobHash(sha1, conn, keyspace))
+
+      (duplicates, similar)
     }
   }
 
-  // TODO: should return something later
-  def findSimilarForFile(file: File, bblfshClient: BblfshClient, feClient: FeatureExtractor): Unit = {
-    val uast = extractUAST(file, bblfshClient)
-    if (uast.isDefined) {
-      log.info(s"uast received: ${uast.toString}")
-      val features = extractFeatures(feClient, uast.get)
-      log.info(s"features: ${features.toString}")
-      // TODO: calculate minhash, find similar
+  def findSimilarForFile(
+                          file: File,
+                          conn: Session,
+                          bblfshClient: BblfshClient,
+                          feClient: FeatureExtractor,
+                          docFreqFile: String,
+                          paramsFile: String,
+                          htnum: Int,
+                          bandSize: Int): Option[Array[String]] = {
+
+    val features = extractUAST(file, bblfshClient).map { uast =>
```

Comment:
You could split this code in smaller private functions. That would help readability.

Reply:
sorry, do you suggest to extract
```
    val features = extractUAST(file, bblfshClient).map { uast =>
      log.debug(s"uast received: ${uast.toString}")

      val result = extractFeatures(feClient, uast)
      log.debug(s"features: ${result.toString}")

      result
    }
```
and the rest in separate functions?

I agree about the rest, but the code above is actually a one-liner but with `log.debug`. I don't see advantages of extraction for it.